test/broadcast.jl

```diff
@@ -164,4 +164,12 @@ end
         @testinf broadcast!(identity, mv, aX) == MVector(aX,aX,aX)
         @test mv == SVector(aX,aX,aX)
     end
+
+    @testset "broadcast! with Array destination" begin
```

Comment:
Can we add a test case where the compiler doesn't know the size of the LHS of `.=` and we have to "expand" to fill it.

E.g. 
```
a = zeros(3,3)
b = SVector(1,2,3)
a .= b
@test a == [1 1 1; 2 2 2; 3 3 3]`
```
or whatever

Reply:
Darn, yes that's broken, good call.

Reply:
Alright, in any case I think the output of `broadcast_size` for both the destination and the arguments should be used to determine (probably in `combine_size` and `broadcast_dest_size`) whether to fall back to generic broadcast code if any of the dimensions is not known at compile time (by changing the `BroadcastStyle`).

Currently, the `broadcast_size` of a non-static `AbstractArray` is just `Size()`, meaning that information about the number of dimensions is lost. So `broadcast_size(::AbstractArray)` should be changed to avoid this information loss.

There are a couple of options:

1. Use `-1` as a sentinel in `Size` for unknown dimensions, e.g. `Size{(-1, -1)}()` to signify the dimensions of a matrix whose dimensions are unknown at compile time.
1. Allow something like `Size{(Unknown(), Unknown())}`, i.e. a more explicit version of 1 (strictly better than 1 in my opinion).
1. Leave `Size` as is and create something like a `BroadcastSize` `struct` that implements 2 above, but keeps changes local to the `broadcast` code. Entails some code duplication, but is probably less likely to break other parts of StaticArrays.
1. Have `broadcast_size(::AbstractArray{<:Any, N}) where {N}` return something like `SizeUnknown{N}()` and make `combine_sizes` able to handle these `SizeUnknown`s as well. Probably the simplest to implement, but perhaps not the most elegant.

Which of these (or some other option) has your preference?

Reply:
Sorry, I've been travelling.

I'm happy for you to use your judgement and play around with one of these options. I'm supposing the easiest way out of this is to simply bail out and call the default `broadcast`(`!`) method whenever the static dimensions are singleton. But you might need to implement one of the above even to correctly detect all such cases. :)